File: packages/crypto/src/contacts.ts
import { CryptoError } from './errors';
import { decryptKey, encryptKey } from './aead';
import { decode, encode } from './keys';

/*
 * Contacts and the settings they are pinned in.
 *
 * A fingerprint (see fingerprint.ts) is what two people compare over another
 * channel before they trust a key the server handed them.
 *
 * Settings (the contact pins, and later anything else that must agree across
 * devices) are one JSON document sealed under a key derived from the identity
 * encryption private key. That key survives password changes, master-key
 * rotation and recovery unchanged, so settings need no rewrap in any of those
 * flows, and they are readable exactly by whoever holds the identity.
 */

export const SETTINGS_MAX_BYTES = 64 * 1024;

export { fingerprint } from './fingerprint';

export type SettingsEnvelope = {
    version: 1;
    /* Bumped by every save; the server refuses a save that names a stale one. */
    settingsVersion: number;
    nonce: string;
    ciphertext: string;
};

export type ContactPin = {
    userId: string;
    email: string;
    name: string;
    encryptionPublicKey: string;
    signingPublicKey: string;
    fingerprint: string;
    pinnedAt: string;
};

export type Settings = {
    version: 1;
    contacts: Record<string, ContactPin>;
};

export const EMPTY_SETTINGS: Settings = { version: 1, contacts: {} };

async function settingsKey(identityPrivateKey: Uint8Array<ArrayBuffer>) {
    const key = await crypto.subtle.importKey('raw', identityPrivateKey, 'HKDF', false, [
        'deriveBits',
    ]);
    return new Uint8Array(
        await crypto.subtle.deriveBits(
            {
                name: 'HKDF',
                hash: 'SHA-256',
                salt: new Uint8Array(32),
                info: new TextEncoder().encode('hushos/settings/v1'),
            },
            key,
            256,
        ),
    ) as Uint8Array<ArrayBuffer>;
}
function settingsContext(userId: string, settingsVersion: number) {
    return new TextEncoder().encode(
        JSON.stringify(['hushos/settings', 1, userId.toLowerCase(), settingsVersion]),
    );
}

export async function sealSettings(
    identityPrivateKey: Uint8Array<ArrayBuffer>,
    userId: string,
    settingsVersion: number,
    settings: Settings,
): Promise<SettingsEnvelope> {
    if (!Number.isSafeInteger(settingsVersion) || settingsVersion < 1)
        throw new CryptoError('Invalid settings version.');
    const plaintext = new TextEncoder().encode(JSON.stringify(settings));
    if (plaintext.length > SETTINGS_MAX_BYTES) throw new CryptoError('Settings are too large.');
    const key = await settingsKey(identityPrivateKey);
    const nonce = crypto.getRandomValues(new Uint8Array(24));
    try {
        return {
            version: 1,
            settingsVersion,
            nonce: encode(nonce),
            ciphertext: encode(
                await encryptKey(plaintext, key, nonce, settingsContext(userId, settingsVersion)),
            ),
        };
    } finally {
        key.fill(0);
        plaintext.fill(0);
    }
}

export async function openSettings(
    identityPrivateKey: Uint8Array<ArrayBuffer>,
    userId: string,
    envelope: SettingsEnvelope,
): Promise<Settings> {
    if (
        envelope.version !== 1 ||
        !Number.isSafeInteger(envelope.settingsVersion) ||
        envelope.settingsVersion < 1
    )
        throw new CryptoError('Unsupported settings envelope.');
    const key = await settingsKey(identityPrivateKey);
    try {
        const plaintext = await decryptKey(
            decode(envelope.ciphertext),
            key,
            decode(envelope.nonce, 24),
            settingsContext(userId, envelope.settingsVersion),
        );
        const parsed: unknown = JSON.parse(new TextDecoder().decode(plaintext));
        plaintext.fill(0);
        if (
            !parsed ||
            typeof parsed !== 'object' ||
            (parsed as Settings).version !== 1 ||
            typeof (parsed as Settings).contacts !== 'object'
        )
            throw new CryptoError('Unsupported settings document.');
        return parsed as Settings;
    } finally {
        key.fill(0);
    }
}

File: crates/hushos-core/src/contacts.rs
//! Contacts: the fingerprint two people compare before trusting a key, and
//! the account settings document that keeps the pins, sealed under a key off
//! the identity's X25519 private key exactly as `packages/crypto/contacts.ts`
//! writes it. The server stores the envelope and a version for compare-and-set.

use crate::bytes::{decode, encode, expect_len, random, KEY_BYTES, NONCE_BYTES};
use crate::envelope::{context, open_with_nonce, seal_with_nonce};
use crate::error::{input, CoreError, CoreResult};
use hkdf::Hkdf;
use serde_json::json;
use sha2::{Digest as _, Sha256};

const SETTINGS_MAX_BYTES: usize = 64 * 1024;

/// The settings document as the server keeps it.
#[derive(Debug, Clone)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]
pub struct SettingsEnvelope {
    pub version: u32,
    pub settings_version: u64,
    pub nonce: String,
    pub ciphertext: String,
}

/// What two people read to each other over another channel: SHA-256 over a
/// domain prefix and the identity's X25519 public key, the first 20 bytes as
/// ten groups of four hex digits.
///
/// ```
/// let print = hushos_core::identity_fingerprint(vec![7u8; 32])?;
/// assert_eq!(print.split(' ').count(), 10);
/// assert!(print.chars().all(|c| c.is_ascii_hexdigit() || c == ' '));
/// # Ok::<(), hushos_core::CoreError>(())
/// ```
#[cfg_attr(feature = "uniffi", uniffi::export)]
pub fn identity_fingerprint(encryption_public_key: Vec<u8>) -> CoreResult<String> {
    expect_len("identity public key", &encryption_public_key, 32)
        .map_err(|_| input("Invalid identity key."))?;
    let mut hasher = Sha256::new();
    hasher.update(b"hushos/identity/fingerprint/v1");
    hasher.update(&encryption_public_key);
    let digest = hasher.finalize();
    let hex: String = digest[..20].iter().map(|b| format!("{b:02x}")).collect();
    Ok(hex.as_bytes().chunks(4).map(|c| std::str::from_utf8(c).expect("hex")).collect::<Vec<_>>().join(" "))
}

/// SHA-256 of a public key as hex: what a pin keeps of a contact's ML-KEM key instead of 1184 bytes.
#[cfg_attr(feature = "uniffi", uniffi::export)]
pub fn key_digest(public_key: Vec<u8>) -> String {
    Sha256::digest(&public_key).iter().map(|b| format!("{b:02x}")).collect()
}

fn settings_key(identity_private_key: &[u8]) -> CoreResult<[u8; KEY_BYTES]> {
    expect_len("identity private key", identity_private_key, 32)
        .map_err(|_| input("Invalid identity key."))?;
    let mut key = [0u8; KEY_BYTES];
    Hkdf::<Sha256>::new(Some(&[0u8; 32]), identity_private_key)
        .expand(b"hushos/settings/v1", &mut key)
        .map_err(|_| input("HKDF failed"))?;
    Ok(key)
}

fn settings_context(user_id: &str, settings_version: u64) -> Vec<u8> {
    context(&[json!("hushos/settings"), json!(1), json!(user_id.to_lowercase()), json!(settings_version)])
}

/// Seals the settings JSON (`{"version":1,"contacts":{...}}`) at the version
/// the caller will send as the next one; the server accepts it only if the
/// stored version is the one the caller read.
///
/// ```
/// use hushos_core::{settings_open, settings_seal};
/// let key = vec![9u8; 32];
/// let sealed = settings_seal(key.clone(), "User-1".into(), 3, r#"{"version":1,"contacts":{}}"#.into())?;
/// assert_eq!(sealed.settings_version, 3);
/// assert_eq!(settings_open(key, "user-1".into(), sealed)?, r#"{"version":1,"contacts":{}}"#);
/// # Ok::<(), hushos_core::CoreError>(())
/// ```
#[cfg_attr(feature = "uniffi", uniffi::export)]
pub fn settings_seal(
    identity_private_key: Vec<u8>,
    user_id: String,
    settings_version: u64,
    json: String,
) -> CoreResult<SettingsEnvelope> {
    if settings_version < 1 {
        return Err(input("Invalid settings version."));
    }
    let parsed: serde_json::Value =
        serde_json::from_str(&json).map_err(|_| input("Settings are not JSON."))?;
    if parsed["version"] != json!(1) || !parsed["contacts"].is_object() {
        return Err(input("Unsupported settings document."));
    }
    if json.len() > SETTINGS_MAX_BYTES {
        return Err(input("Settings are too large."));
    }
    let key = settings_key(&identity_private_key)?;
    let nonce = random(NONCE_BYTES);
    let sealed =
        seal_with_nonce(&key, &nonce, json.as_bytes(), &settings_context(&user_id, settings_version))?;
    Ok(SettingsEnvelope { version: 1, settings_version, nonce: encode(&nonce), ciphertext: encode(&sealed) })
}

/// The settings JSON out of its envelope.
#[cfg_attr(feature = "uniffi", uniffi::export)]
pub fn settings_open(
    identity_private_key: Vec<u8>,
    user_id: String,
    envelope: SettingsEnvelope,
) -> CoreResult<String> {
    if envelope.version != 1 || envelope.settings_version < 1 {
        return Err(input("Unsupported settings envelope."));
    }
    let key = settings_key(&identity_private_key)?;
    let plaintext = open_with_nonce(
        &key,
        &decode("nonce", &envelope.nonce, Some(NONCE_BYTES))?,
        &decode("ciphertext", &envelope.ciphertext, None)?,
        &settings_context(&user_id, envelope.settings_version),
    )?;
    let text = String::from_utf8(plaintext)
        .map_err(|_| CoreError::Sealed("Unsupported settings document.".into()))?;
    let parsed: serde_json::Value = serde_json::from_str(&text)
        .map_err(|_| CoreError::Sealed("Unsupported settings document.".into()))?;
    if parsed["version"] != json!(1) || !parsed["contacts"].is_object() {
        return Err(CoreError::Sealed("Unsupported settings document.".into()));
    }
    Ok(text)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_fingerprint_is_the_web_layout_and_changes_with_the_key() {
        let a = identity_fingerprint(vec![1u8; 32]).unwrap();
        let b = identity_fingerprint(vec![2u8; 32]).unwrap();
        assert_eq!(a.len(), 49, "ten groups of four and nine spaces");
        assert_ne!(a, b);
        assert!(matches!(identity_fingerprint(vec![1u8; 31]), Err(CoreError::Input(_))));
        assert_eq!(key_digest(vec![]).len(), 64);
    }

    #[test]
    fn settings_open_only_for_the_same_identity_user_and_version() {
        let key = random(32);
        let sealed =
            settings_seal(key.clone(), "u".into(), 2, r#"{"version":1,"contacts":{"x":{}}}"#.into()).unwrap();
        assert!(matches!(settings_open(random(32), "u".into(), sealed.clone()), Err(CoreError::Sealed(_))));
        assert!(matches!(settings_open(key.clone(), "v".into(), sealed.clone()), Err(CoreError::Sealed(_))));
        let mut bumped = sealed.clone();
        bumped.settings_version = 3;
        assert!(matches!(settings_open(key.clone(), "u".into(), bumped), Err(CoreError::Sealed(_))));
        assert_eq!(settings_open(key, "U".into(), sealed).unwrap(), r#"{"version":1,"contacts":{"x":{}}}"#);
    }

    #[test]
    fn malformed_or_oversized_settings_are_refused_before_sealing() {
        let key = random(32);
        assert!(matches!(
            settings_seal(key.clone(), "u".into(), 1, "not json".into()),
            Err(CoreError::Input(_))
        ));
        assert!(matches!(
            settings_seal(key.clone(), "u".into(), 1, r#"{"version":2,"contacts":{}}"#.into()),
            Err(CoreError::Input(_))
        ));
        assert!(matches!(
            settings_seal(key.clone(), "u".into(), 0, r#"{"version":1,"contacts":{}}"#.into()),
            Err(CoreError::Input(_))
        ));
        let huge = format!(r#"{{"version":1,"contacts":{{"a":"{}"}}}}"#, "x".repeat(SETTINGS_MAX_BYTES));
        assert!(matches!(settings_seal(key, "u".into(), 1, huge), Err(CoreError::Input(_))));
    }
}
